### ISSUE TITLE
self conditioning not working (size mismatch)

### ISSUE THREAD

**Issue by inspirit** [lucidrains/soundstorm-pytorch#35]
When enabling self_cond there is size mismatch when trying to sum embeds: 
https://github.com/lucidrains/soundstorm-pytorch/blob/08a5e20c322f69350a2ed1abc2d7144b463e8d13/soundstorm_pytorch/soundstorm.py#L582

this happens because 'sum_embeds' tensor is not reduced as it was projected to:
```
nn.Linear(dim, dim * num_effective_quantizers),
Rearrange('b n (h d) -> b (n h) d', h = num_effective_quantizers)
```
so 'sum_embeds' is actually 'b (n h) d' while 'x' is 'b n d'

**Comment by lucidrains** [comment]
@inspirit ah, thanks for uncovering this

how about 0.4.4?

**Comment by inspirit** [comment]
you forgot to put 'sum' argument to 'reduce' call :)

**Comment by inspirit** [comment]
ohh, and btw there is a special case when 'generate' is called because you will get 'null_embed' at the first iteration which is just [dim] shape/size

**Comment by inspirit** [comment]
probably smth like:
```
if sum_embeds.ndim == 3:
    sum_embeds = reduce(sum_embeds, 'b (n h) d -> b n d', 'sum', h = self.num_effective_quantizers)
```
should do the trick

**Comment by lucidrains** [comment]
@inspirit ok done

does self conditioning work for you? haven't heard of many people using it

**Comment by inspirit** [comment]
i just started digging it for one of my projects with Residual Quantizers, so too early to comment, was validating that all works after incorporating it to my pipeline and found this issue, so might be completely useless in the end :)
